Refactored tests and added new tests tools for videos-tests.

huguesdevimeux opened the pull request:
Ye so as you may have noticed tests are poor. They are not well organized, very hard to debug, and don't cover enough things (only frames comparison and some logging stuff) :/ 

**NOTE: This PR is only meant to add new tests tools, not new tests. For two reasons : This pr would be too big, and I'm getting quite super sick of tests lol.** 

#  List of Changes (sorted by importance) (UP TO DATE) : 
- Refactored directories organisation: Now it looks like the one said in #227 . (note that it is temporary, I will change it a little bit later. 
- Added --skip_slow flag for pytest, that would skip the slow tests. 
- Added a show_diff flag for pytest, that visually display differences in graphical units tests if one fails. Example : 
![Figure_1](https://user-images.githubusercontent.com/36239975/90978823-ecba5400-e550-11ea-94ad-83b4bc375ea6.png)
- Added tools to test things involving scene-rendering (eg, flags etc), with `@video_comparison` decorator. Example  
```python 
@video_comparison(
    "SquareToCircleWithPFlag.json", "videos/simple_scenes/480p15/SquareToCircle.mp4"
)
def test_basic_scene_p_flag(tmp_path, manim_cfg_file, simple_scenes_path):
    scene_name = "SquareToCircle"
    command = [
        "python",
        "-m",
        "manim",
        simple_scenes_path,
        scene_name,
        "-l",
        "--media_dir",
        str(tmp_path),
    ]
    out, err, exit_code = capture(command)
    assert exit_code == 0, err
```
Read below the explanations

-  Added helpers to set new tests. 
- Better use of pytest syntax for graphical units tests. We now use `@pytest.mark.parametrize` for them.
- Now the tests use tempfiles. It means that they have to generate cache everytime, which make them a lot slower. 
- Miscellaneous docs, comments, etc improvements. 


## Explanation for Changes
So, videos tests work as the following : 
We basically test two things, as indicated in #227 : video config (eg, fps, duration) and hash of the video. To get them, we use `ffprobe` and `ffpemg` (ffprobe is installed with ffpemg). Then, with the helper `save_control_data_from_video` in helpers/video_utils.py, we save this information in a .json format in control_data/videos_data that we will use to compare the new information extracted from the video to test. 

## Testing Status

Wait, it's all a test of tests? 

Always have been. 

## Further Comments
I don't think I will add utils to loggingtests. On second thought, I don't find them useful and they are not worth it. Please tell me your opinion.

## Acknowledgement
- [x] I have read the [Contributing Guidelines](https://github.com/ManimCommunity/manim/wiki/Documentation-guidelines-(WIP))

leotrs commented on tests/conftest.py:
```diff
@@ -4,15 +4,20 @@
 def pytest_addoption(parser):
     parser.addoption(
-        "--skip_end_to_end",
+        "--skip_slow",
```
Question: can you please explain this flag? Does it always skip all tests that are slow? If so, how/when do we test the slow ones?

huguesdevimeux replied:
I will document this ASAP !

leotrs commented on tests/conftest.py:
```diff
@@ -4,15 +4,20 @@
+        "--show_diff",
+        action="store_true",
+        default=False,
+        help="Will show a vizual comparison if a graphical unit test fails.",
```
```suggestion
        help="Will show a visual comparison if a graphical unit test fails.",
```

leotrs commented on tests/test_scene_rendering/simple_scenes.py:
```diff
@@ -0,0 +1,8 @@
+from manim import *
+
+
+class SquareToCircle(Scene):
+    def construct(self):
+        square = Square()
+        circle = Circle()
+        self.play(Transform(square, circle))
```
Question: why include only one of the example scenes?

huguesdevimeux replied:
For the moment only one is needed, as I use it only for flags-testing.

leotrs replied:
I think we should have all of them in the future, as they use different funcionalities.

huguesdevimeux replied:
Of course. The system is made to completed in the future with new scenes, etc. It's totally flexible

leotrs commented on tests/test_scene_rendering/conftest.py:
```diff
@@ -0,0 +1,17 @@
+from tempfile import TemporaryDirectory
+
+
+from manim import file_writer_config
```
Nitpick: unused import?

huguesdevimeux replied:
Good catch. Thanks

leotrs commented on tests/test_scene_rendering/test_scene.py:
```diff
@@ -0,0 +1,49 @@
+    ]
+    out, err, exit_code = capture(command)
+    assert exit_code == 0, err
+    # print(out)
+    # print(out)
+    # assert os.path.isfile()
```
```suggestion
```

leotrs commented on tests/test_scene_rendering/test_scene.py:
```diff
@@ -0,0 +1,49 @@
+)
+def test_basic_scene_p_flag(tmp_path, manim_cfg_file, simple_scenes_path):
+    scene_name = "SquareToCircle"
+    command = [
+        "python",
+        "-m",
+        "manim",
+        simple_scenes_path,
+        scene_name,
+        "-l",
+        "--media_dir",
+        str(tmp_path),
+    ]
```
Issue: this doesn't seem to be using the -p flag.

Aathish04 replied:
The `-p` flag won't work on headless machines like the Github Actions runners, since they don't have video players.

leotrs replied:
Right, that makes sense. But I thought this test was supposed to be testing the -p flag. It's named `test_basic_scene_p_flag` after all.

huguesdevimeux replied:
yes, it's my fault, it's a typo. By `-p` flag I meant `-l` flag.

leotrs commented on tests/helpers/graphical_units.py:
```diff
@@ -0,0 +1,50 @@
+import os
```
Comment: it's unclear to me what's the difference between tests/helpers and tests/utils.

huguesdevimeux replied:
`helpers `is meant to be used to set new tests up (i.e, generate control data as prerendered .npy, or pre-rendered videos hashes/config), when utils is used by pytest. 

Maybe the wording is not good. Do you have a suggestion?

leotrs replied:
Oh ok. I'd suggest adding a module-level docstring to these two files that specify which one is for the user/dev and which one is for pytest.

On top of that, I'd suggest changing their names to something like dev_tools and pytest_tools or something to that effect. But this may be overkill :shrug:

huguesdevimeux replied:
> Oh ok. I'd suggest adding a module-level docstring to these two files that specify which one is for the user/dev and which one is for pytest.

How . By creating a README ?

leotrs replied:
No, by literally adding a docstring at the top of each file.

huguesdevimeux replied:
Oh, of course. Will do that

leotrs commented on tests/helpers/graphical_units.py:
```diff
@@ -0,0 +1,50 @@
+
+def set_test_scene(scene_object, module_name):
+    """Function used to set up the test data for a new feature. This will basically set up a pre-rendered frame for a scene. This is meant to be used only
+    when setting up tests. Please refer to the wiki.
```
Suggestion: add wiki URL.

huguesdevimeux replied:
Will do when I will rewrite the wiki !

huguesdevimeux commented:
@leotrs 
> Question: I wasn't able to find a file testing logging. Before we had test_logging/test_logging, but that file seems to have been removed.

I haven't merged from master now. I will incorporate them when I do so.

huguesdevimeux commented:
Could someone check if tests work locally ? They are working fine on mine.

Aathish04 commented:
> Could someone check if tests work locally ? They are working fine on mine.

The tests fail on my system.

The failing tests were the last two tests in `tests/test_graphical_units/test_writing.py` and `tests/test_logging/test_logging.py`.

Not sure why test_writing fails, but apparently it doesn't match with the reference data.
`test_logging` fails because, as far as I can tell, the log file wasn't created in the right place (or at all, actually).

huguesdevimeux commented:
@Aathish04 I tried several times to put the old control data for the writing tests, without success. Could you try to re-generate some on your computer? I would be grateful.

> Not sure why test_writing fails, but apparently it doesn't match with the reference data.
> test_logging fails because, as far as I can tell, the log file wasn't created in the right place (or at all, actually).

And uuuuh ... last favor .. Could you try to investigate on this ? I can't, as tests are working fine on my computer. 

When tests will be fixed this PR will be ready to merge !

Aathish04 commented:
> @Aathish04 I tried several times to put the old control data for the writing tests, without success. Could you try to re-generate some on your computer? I would be grateful.

Ok, so _this_ test fails, and will continue to fail, as long as we keep single reference data for all OSes. You can see in the most recent commit that the test for writing  passed in MacOS (where I generated the reference data), but fails on the other operating systems.

Do you know how we fixed this _before_ this refactoring? I can't remember how we made the writing tests pass for all the OSes...



> > test_logging fails because, as far as I can tell, the log file wasn't created in the right place (or at all, actually).

Still trying to figure out where the tempfile is created... can't seem to find it.

huguesdevimeux commented:
> Ok, so _this_ test fails, and will continue to fail, as long as we keep single reference data for all OSes. You can see in the most recent commit that the test for writing passed in MacOS (where I generated the reference data), but fails on the other operating systems.
> 
> Do you know how we fixed this _before_ this refactoring? I can't remember how we made the writing tests pass for all the OSes...

This is the issue, I didn't change anything about these tests, I just moved dirs and codes. But notice, only the last test_writing is failing now! it's better! (but it still fails :( ) If no fix is found by either me or you, I will mark this test as xfail (I.e if its fails, it does not mark the whole tests as failed. 

> Still trying to figure out where the tempfile is created... can't seem to find it.

The logging_test only fails on macos. The tempfile is generated by pytest somewhere over the system, (for example, /dev/temp). I don't think the tempdir is the issue here, maybe the issue is the files generated within it (i.e, `media/logs/` etc). Maybe a permission issue ?

Aathish04 commented:
> The logging_test only fails on macos. The tempfile is generated by pytest somewhere over the system, (for example, /dev/temp). I don't think the tempdir is the issue here, maybe the issue is the files generated within it (i.e, `media/logs/` etc). Maybe a permission issue ?

Did some more poking around, and out of desperation, I tried to use shell mode in `subprocess.popen` for the test. The logging test works in MacOS when shell mode is used. I've committed that change, and fingers crossed, hopefully it won't break any other test...

Aathish04 commented:
@huguesdevimeux I can safely say the `test_logging` issue has been fixed by using shell mode in `subprocess.popen`.

All the Tests pass on MacOS, and only  `test_writing` fails on Ubuntu and for some reason both `test_writing` and `test_scene` fail on windows with python3.8 but `test_scene` passes in the other ones...

I think the only thing left to do is to fix the latex tests and find why `test_scene` fails and then we'll be ready for review from the others.

huguesdevimeux commented:
I removed hash tests from videos tests. Indeed, testing video content appears to be impossible, as ffmpeg is too much os-dependent. See https://stackoverflow.com/questions/46898384/ffmpeg-libx264-encoder-different-machines-different-output

huguesdevimeux commented:
Review at will !

leotrs commented on tests/test_logging/test_logging.py:
```diff
@@ -5,37 +5,47 @@
+    path_basic_scene = os.path.join("tests", "test_logging", "basic_scenes.py")
+    path_output = os.path.join(tmp_path, "media_temp")
+    os.makedirs(tmp_path, exist_ok=True)
+    command = " ".join(
```
Question: why is this command being `join`ed while the command in `tests/test_scene_rendering/test_cli_flags.py::L17` isn't?

Aathish04 replied:
This is because this test sends the command to `subprocess.popen` to execute using `shell mode`, which requires that the command be sent as a string.

The reason it has to be sent as a string is due to some quirks of MacOS, which I did not delve too deep into. In short, if we don't send it over as a string, this test's temporary directories are not created, causing the test to fail. I _think_ this is because the permissions are set differently in MacOS, preventing pytest (via manim) from writing to the temporary media directories.

We don't send it over as a string in `tests/test_scene_rendering/test_cli_flags.py::L17`, because the same issue was not faced there, meaning the temporary directories were created and recognised by MacOS.

leotrs replied:
:woozy_face:

leotrs commented on tests/utils/GraphicalUnitTester.py:
```diff
@@ -25,38 +26,59 @@ class SceneTester:
         config["pixel_width"] = 854
         config["frame_rate"] = 15
+
+        for dir_temp in [
```
Question: is there any reason we are not using [tempfile](https://docs.python.org/3/library/tempfile.html)? If at all possible, we should definitely use it. There's also [pytest tempdir](https://docs.pytest.org/en/stable/tmpdir.html) which is perhaps more appropriate.

If it is possible to use either of these, we should leave that for a later PR.

leotrs replied:
Oh wait I see I think the `tmpdir` argument is already a folder created by tempfile, no?

huguesdevimeux replied:
It already used a `tempdir`, see l.45 of this file. 

>  think the tmpdir argument is already a folder created by tempfile, no?

Yep (but in fact, created by `pytest`)

leotrs commented on tests/utils/video_tester.py:
```diff
@@ -0,0 +1,84 @@
+            tests_directory = os.path.dirname(
+                os.path.dirname(os.path.abspath(__file__))
+            )
+            path_control_data = os.path.join(
+                tests_directory, "control_data", "videos_data", control_data_file
+            )
+            path_video_generated = tmp_path / scene_path_from_media_dir
```
Question: correct me if I'm wrong, but `os.path.join` uses strings to represent paths, and but L71 here is using `Path` objects, no? This seems messy.

leotrs replied:
(This is messy, but non-blocking)

huguesdevimeux replied:
I see. I will try to change this.

huguesdevimeux replied:
On second thought I will change it later when i will  changeall the file handling in the tests.

huguesdevimeux commented on tests/utils/GraphicalUnitTester.py:
```diff
@@ -25,38 +26,59 @@ class SceneTester:
     path_tests_medias_cache : : class:`str`
         Path to 'media' folder generated by manim. This folder contains cached data used by some tests.
-    path_tests_data : : class:`str`
+    path_control_data : : class:`str`
```
Wooop very good catch. Will change that !

huguesdevimeux replied:
**UPDATE :** It's right, as this is the list of the arguments, not the parameters. You didn't see it because Github hide it.

leotrs replied:
lol

huguesdevimeux replied:
No really its in the attributes list it was "wrapped" by github

PgBiel commented on tests/helpers/graphical_units.py:
```diff
@@ -0,0 +1,54 @@
+"""Helpers functions for devs to set up new graphical-units data.  
+"""
```
unnecessary line break
```suggestion
"""Helpers functions for devs to set up new graphical-units data."""
```

PgBiel commented on tests/helpers/graphical_units.py:
```diff
@@ -0,0 +1,54 @@
+
+    Examples
+    --------
+    Normal usage::
+        set_test_scene(DotTest, "geometry")
```
```suggestion
    Normal usage::
    
        set_test_scene(DotTest, "geometry")
```

PgBiel commented on tests/helpers/video_utils.py:
```diff
@@ -0,0 +1,61 @@
+"""Helpers for dev to set up new tests that use videos. 
+"""
```
```suggestion
"""Helpers for dev to set up new tests that use videos."""
```

PgBiel commented on tests/helpers/video_utils.py:
```diff
@@ -0,0 +1,61 @@
+    """Helper used to set up a new test that will compare videos. This will create a new .json file in control_data/videos_data that contains
+    informations tested of the video, including its hash. Refer to the wiki for more informations.
+
+    Parameters :
+    ------------
```
```suggestion
    Parameters
    ----------
```

PgBiel commented on tests/helpers/video_utils.py:
```diff
@@ -0,0 +1,61 @@
+    Parameters :
+    ------------
+    path_to_video : :class:`str`
+        Path to the video to extract  informations from.
```
```suggestion
        Path to the video to extract information from.
```

PgBiel commented on tests/helpers/video_utils.py:
```diff
@@ -0,0 +1,61 @@
+        Path to the video to extract  informations from.
+    name : :class:`str`
+        Name of the test. The .json file will be named with it.
+
```
```suggestion
```

PgBiel commented on tests/utils/testing_utils.py:
```diff
@@ -0,0 +1,21 @@
+
+    Returns
+    -------
+    :class:`list`
```
unspecific types so generic type; gotta specify type of list
```suggestion
    List[:class:`type`]
```

PgBiel commented on tests/utils/video_tester.py:
```diff
@@ -0,0 +1,84 @@
+def video_comparison(control_data_file, scene_path_from_media_dir):
+    """Decorator used for any test that needs to check a rendered scene/video.
+
+    Parameters:
+    -----------
```
no colon
```suggestion
    Parameters
    ----------
```

huguesdevimeux commented:
@PgBiel I adressed all your changes !

PgBiel commented on tests/utils/GraphicalUnitTester.py:
```diff
@@ -25,38 +26,59 @@ class SceneTester:
     path_tests_medias_cache : : class:`str`
         Path to 'media' folder generated by manim. This folder contains cached data used by some tests.
-    path_tests_data : : class:`str`
+    path_control_data : : class:`str`
```
```suggestion
    path_control_data : :class:`str`
```

PgBiel commented on tests/utils/video_tester.py:
```diff
@@ -0,0 +1,84 @@
+    Parameters
+    ----------
+    control_data_file : :class:`str`
+        Name of the control data file, ie the .json containing all the pre-rendered references of the scene tested. Warning : you don't have to pass the path here.
```
Might be better to have it as a directive here
```suggestion
        Name of the control data file, i.e. the .json containing all the pre-rendered references of the scene tested.
        .. warning:: You don't have to pass the path here.
```
